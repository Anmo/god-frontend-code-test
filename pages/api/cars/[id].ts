import type { NextApiRequest, NextApiResponse } from "next";

import cars from "../../../public/api/cars.json";
import type { Car } from "../../../src/types";

export default function handler(
  req: NextApiRequest,
  res: NextApiResponse<Car | { error: string }>
) {
  const car = cars.find(({ id }) => id === req.query.id);

  if (!car) {
    return res.status(404).json({ error: "Car not found" });
  }

  res.status(200).json(car);
}
