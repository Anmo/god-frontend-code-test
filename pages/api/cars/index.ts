import type { NextApiRequest, NextApiResponse } from "next";

import cars from "../../../public/api/cars.json";
import type { CarBodyType, Cars } from "../../../src/types";

export default function handler(
  req: NextApiRequest,
  res: NextApiResponse<Cars>
) {
  const bodyType = req.query.bodyType as CarBodyType | undefined;
  return res
    .status(200)
    .json(bodyType ? cars.filter((car) => car.bodyType === bodyType) : cars);
}
