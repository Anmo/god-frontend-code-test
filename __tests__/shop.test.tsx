import { vi, describe, expect, it, beforeEach, Mock, afterEach } from "vitest";
import { render, screen } from "@testing-library/react";
import { createDynamicRouteParser } from "next-router-mock/dynamic-routes";
import mockRouter from "next-router-mock";
import useSWR from "swr";

import cars from "../public/api/cars.json";
import Shop from "../pages/shop/[id]";

vi.mock("swr");
const useSWRMock = useSWR as Mock;

mockRouter.useParser(createDynamicRouteParser(["/shop/[id]"]));

describe("Shop", () => {
  beforeEach(() => {
    mockRouter.push("shop/1");
    useSWRMock.mockReturnValue({});
  });

  afterEach(() => {
    useSWRMock.mockClear();
  });

  it("renders loading state", () => {
    // When
    render(<Shop />);

    // Then
    expect(screen.getByText("Loading...")).toBeDefined();
  });

  it("renders error state", () => {
    // Given
    useSWRMock.mockReturnValueOnce({
      error: "Something wrong is not correct!",
    });

    // When
    render(<Shop />);

    // Then
    expect(screen.getByText("Failed to load")).toBeDefined();
  });

  it("fetches the car with the correct id", () => {
    // When
    render(<Shop />);

    // Then
    expect(useSWRMock).toBeCalledWith("/api/cars/1", expect.any(Function));
  });

  it("renders error state", () => {
    // Given
    useSWRMock.mockReturnValueOnce({ data: cars[0] });

    // When
    const { container } = render(<Shop />);

    // Then
    expect(container).matchSnapshot();
  });
});
