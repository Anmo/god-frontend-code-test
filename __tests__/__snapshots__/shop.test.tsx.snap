// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shop > renders error state 1`] = `
<div>
  <div
    class="px-pagemargin self-center"
  >
    <div
      class="flex-col flex-shrink-0 flex-grow-0 gap-y-8 w-xs h-full snap-start"
    >
      <p
        class="font-medium"
      >
        suv
      </p>
      <div
        class="flex-col md:flex-row gap-8"
      >
        <p
          class="font-medium"
        >
          XC90 Recharge
        </p>
        <p
          class="font-medium text-secondary"
        >
          plug-in hybrid
        </p>
      </div>
      <img
        alt="XC90 Recharge"
        class="cover aspect-4/3 w-full mb-8"
        data-nimg="1"
        decoding="async"
        height="300"
        loading="lazy"
        src="/_next/image?url=%2Fimages%2Fxc90_recharge.jpg&w=828&q=75"
        srcset="/_next/image?url=%2Fimages%2Fxc90_recharge.jpg&w=640&q=75 1x, /_next/image?url=%2Fimages%2Fxc90_recharge.jpg&w=828&q=75 2x"
        style="color: transparent; height: auto;"
        width="400"
      />
      <div
        class="flex flex-wrap gap-x-24 justify-center"
      >
        <a
          class="text-accent-blue"
          href="/"
        >
          <img
            class="icon-mask"
            height="16"
            role="presentation"
            src="https://www.volvocars.com/static/shared/icons/v2/navigation-chevronback-16.svg"
            style="--icon-url: url(https://www.volvocars.com/static/shared/icons/v2/navigation-chevronback-16.svg);"
            width="16"
          />
           Back
        </a>
        <button
          class="text-accent-blue"
          disabled=""
        >
          BUY
        </button>
      </div>
    </div>
  </div>
</div>
`;
